utilities/include/alps/numeric/tensors/mpi/mpi_shared_allocator.hpp

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright (C) 1998-2018 ALPS Collaboration. See COPYRIGHT.TXT
+ * All rights reserved. Use is subject to license terms. See LICENSE.TXT
+ * For use in publications, see ACKNOWLEDGE.TXT
+ */
+
+#ifndef ALPSCORE_MPI_SHARED_ALLOCATOR_HPP
+#define ALPSCORE_MPI_SHARED_ALLOCATOR_HPP
+
+#include <alps/config.hpp>
+
+#ifdef ALPS_HAVE_MPI
+#include <mpi.h>
+
+#if MPI_VERSION > 2
+#define ALPS_HAVE_SHARED_ALLOCATOR
+#else
+#ifndef NDEBUG
+#warning "MPI shared memory is not supported by your MPI library."
+#endif
+#endif
+
+#ifdef ALPS_HAVE_SHARED_ALLOCATOR
+namespace alps {
+  namespace numerics {
+    namespace detail {
+      template <typename T>
+      class mpi_shared_allocator {
+      public:
+
+        mpi_shared_allocator(MPI_Comm& comm, MPI_Win& win) : comm_(&comm), win_(&win) {
+
+        };
+
+        T* allocate(size_t size) {
+          void *mem;
+          // compute local size for array
+          size_t nloc = local_size(size);
+          // allocate shared memory
+          MPI_Win_allocate_shared(nloc * sizeof(T), sizeof(T), MPI_INFO_NULL, *comm_, &mem, win_);
+          MPI_Aint alloc_length;
+          int disp_unit;
+          // get pointer to the 0-th element of the global array
+          MPI_Win_shared_query(*win_, 0, &alloc_length, &disp_unit, &mem);
+          return (T*) mem;
+        }
+
+        void deallocate(T* data) {
+          // we do not release anything.
+          // one should explicitly release MPI-window in the very original place.
+        }
+        void lock(){
+          lock_ = true;
+          MPI_Win_lock_all(MPI_MODE_NOCHECK , *win_);
+        }
+        void release(){
+          MPI_Win_sync(*win_);
+          MPI_Win_unlock_all(*win_);
+          lock_ = false;
+        }
```

Comment:
Hang on, isn't that allocator locking model broken by design?

Suppose you have two tensors with two shared_allocators, say `a` and `b`, so you will get two allocators where the locks are bound to the same window, in other words, if  allocator `b` locks, then `b.lock_ = false`, even though the underlying window is locked, isn't it?

Reply:
In other word. When you trying to use MPI, you should think properly, and understand what you are doing. 

Saying "broken by design" you can say simply that MPI is broken by design. Which is not, you just need to properly design your algorithm.

Reply:
No, I'm sorry but you cannot blame this on MPI alone.  This class is just not well-designed.  This code breaks:

```C++
  mpi_shared_allocator a(...);
  mpi_shared_allocator b = a;
  b.lock();
  // a is not locked; i.e., use a -> boom.
```

which also means you cannot embed the allocator into a class, which you can usually do:

```C++
template <T, Alloc>
class my_vector {

private:
     T *begin, *end, *capacity;
     Alloc something;  // put in mpi_shared_allocator -> boom.
```

Allocations leak memory:

```C++
  {
    MPI_Win win;
    ...
    mpi_shared_allocator a(win);
    // gets new window which you dont have access to
    T *p1 = a.allocate(10);
    a.deallocate(p1);
  }  // a is destructor -> leaks internal window -> boom.
```

Reply:
```C++
  mpi_shared_allocator a(...);
  mpi_shared_allocator b = a;
  b.lock();
  // a is not locked; i.e., use a -> boom.
```

as long as you read it - there is nothing critical. You wil probably not get the most recent RMA updates since your cached data is wrong. But this is again, I can do exactly the same with pure MPI without any encapsulation.

> which also means you cannot embed the allocator into a class, which you can usually do:

```C++
template <T, Alloc>
class my_vector {

private:
     T *begin, *end, *capacity;
     Alloc something;  // put in mpi_shared_allocator -> boom.
```

You still can do this if you do proper copy-construction of allocator.

> Allocations leak memory:

no they don't, you should manage memory with window. That is the reason to have window outside. It's the way to have proper copy-semantic. You can't do any copy-construction of shared object If you free memory within allocator destructor. Other way is to disallow copying of shared allocator, so you can't assign or pass by value shared object. 

I don't know which one is better. The second seems easy to implement, but it's kinda not flexible.

BTW, I will remove shared-memory from this PR, don't have time for it right now.

Reply:
Again, either I am really, really missing something, or I think it leaks memory:

```C++
    MPI_Win *win = nullptr;
    {
        mpi_shared_allocator alloc(MPI_COMM_WORLD, win);

        // what happens in allocate():  you call `MPI_Win_allocate_shared`, which creates a
        // new window and overrides the `win_` variable with it. (you are still fine because
        // you have a handle to the old win_ from outside.)
        T *p1 = alloc.allocate(10);
        alloc.deallocate(p1);

        // Now you're screwed.  You just overrode the window with a new window without
        // calling MPI_Free_Win on the current one first.  This will usually happen 
        // somewhere deep in your code (e.g., a vector allocating space for more elements).
        T *p2 = alloc.allocate(10);
        alloc.deallocate(p2);
    }

    // At this point you destroy the allocator, but you DO NOT call MPI_Free_Win on the
    // first one which means the new internal window does not get destroyed.  You 
    // cannot destroy it yourself, because alloc does not give you access to it.
    if (win != nullptr)
        MPI_Free_Win(win);
```

Reply:
Here is another, maybe less contrived, example:

```C++
struct mydata {
    mydata(mpi_shared_allocator<int> alloc) : alloc(alloc), data(alloc.allocate(10)) { }
    mydata(const mydataset &other) : mydataset(other) { }
    ~mydata() { alloc.deallocate(data); }    
private:
    mydata &operator=(const mydata &other) = delete;
    int *data;
    mpi_shared_allocator<int> alloc;
};

MPI_Win *win = nullptr;
mydata x(mpi_shared_allocator<int>(win));
mydata y = x;   // Boom: leaks old window, no way of fixing
if (win != nullptr)
        MPI_Win_free(*win);
```

Reply:
Hey-hey-hey, careful.

You can not do this:

```C++
    MPI_Win *win = nullptr;
```

it should be

```C++
    MPI_Win win;
```

and then something like this.

```C++
    if (win != MPI_WIN_NULL)
        MPI_Win_free(&win);
```

Reply:
OK, let's agree to disagree.  If we can't agree on whether those two examples point to an API that is fundamentally broken versus just one where one needs to understand what one is doing, we won't reach common ground here.